## Overview
Clean BenefitCostSharing by Handle missing value and Sepatate file between $, numeric and string

### Objectives
1. Loading and inspecting dataset
2. Select particular BusinessYear 2014
3. Select only the columns that you want to analyze
4. Separate the string to find groups of string
5. Information used in Two-way ANOVA
6. Handle this data
7. Find out Mean of Copay
8. Drop Duplicates values

### 1. Loading and inspecting dataset
#### import library
1. pandas that it allows you to create, manipulate, and analyze datasets efficiently.
2. numpy that it provides support for arrays, matrices, and various mathematical functions.
3. seaborn that it provides high-level functions to create attractive and informative plots.
4. matplotlib is a foundational library for creating static, interactive, and animated visualizations.

In [40]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [41]:
# Load the dataset
df_bcs = pd.read_csv("BenefitsCostSharing.csv",low_memory=False)

In [42]:
# inspect the dataset
df_bcs.shape

(5048408, 32)

In [43]:
# Countns BusinessYear
df_bcs['BusinessYear'].value_counts().reset_index()

BusinessYear    count
0          2015  2079286
1          2016  1804253
2          2014  1164869

### 2. Select particular BusinessYear 2014
1. Accesses the column BusinessYear from the DataFrame df_bcs2014
3. Counts the occurrences of each unique value in the BusinessYear column
4. Resets the index of the Series so that the unique values and their counts become columns in the new DataFrame

In [44]:
# Keep BusinessYear 2014 in df_bcs2014
df_bcs2014 = df_bcs[df_bcs['BusinessYear'] == 2014]

In [45]:
# 1,164,869 records
df_bcs2014['BusinessYear'].value_counts().reset_index()

BusinessYear    count
0          2014  1164869

In [46]:
## display all columns
pd.set_option('display.max_columns',None)
# display first 5 rows
df_bcs2014.head()

BenefitName  BusinessYear CoinsInnTier1 CoinsInnTier2  \
0  Routine Dental Services (Adult)          2014           20%           NaN   
1     Dental Check-Up for Children          2014           20%           NaN   
2        Basic Dental Care - Child          2014           40%           NaN   
3              Orthodontia - Child          2014           50%           NaN   
4        Major Dental Care - Child          2014           50%           NaN   

  CoinsOutofNet CopayInnTier1 CopayInnTier2 CopayOutofNet  \
0           20%     No Charge           NaN     No Charge   
1           20%     No Charge           NaN     No Charge   
2           40%     No Charge           NaN     No Charge   
3           50%     No Charge           NaN     No Charge   
4           50%     No Charge           NaN     No Charge   

             EHBVarReason Exclusions  \
0               Above EHB        NaN   
1     Substantially Equal        NaN   
2     Substantially Equal        NaN   
3  Additional EHB Benefit        NaN   
4     Substantially Equal        NaN   

                                         Explanation           ImportDate  \
0  Combined annual benefit maximum of $1000 per y...  2014-03-19 07:06:49   
1              See policy for additional limitations  2014-03-19 07:06:49   
2              See policy for additional limitations  2014-03-19 07:06:49   
3  24 month waiting period, See policy for additi...  2014-03-19 07:06:49   
4              See policy for additional limitations  2014-03-19 07:06:49   

  IsCovered IsEHB IsExclFromInnMOOP IsExclFromOonMOOP IsStateMandate  \
0   Covered   NaN                No                No            NaN   
1   Covered   Yes                No                No            NaN   
2   Covered   Yes                No                No            NaN   
3   Covered   NaN                No                No            NaN   
4   Covered   Yes                No                No            NaN   

  IsSubjToDedTier1 IsSubjToDedTier2  IssuerId  IssuerId2  LimitQty  \
0               No               No     21989      21989    1000.0   
1               No               No     21989      21989       1.0   
2              Yes              Yes     21989      21989       NaN   
3              Yes              Yes     21989      21989       NaN   
4              Yes              Yes     21989      21989       NaN   

               LimitUnit  MinimumStay             PlanId QuantLimitOnSvc  \
0       Dollars per Year          NaN  21989AK0010001-00             Yes   
1  Visit(s) per 6 Months          NaN  21989AK0010001-00             Yes   
2                    NaN          NaN  21989AK0010001-00             NaN   
3                    NaN          NaN  21989AK0010001-00             NaN   
4                    NaN          NaN  21989AK0010001-00             NaN   

   RowNumber SourceName StandardComponentId StateCode StateCode2  VersionNum  
0         68       HIOS      21989AK0010001        AK         AK           6  
1        104       HIOS      21989AK0010001        AK         AK           6  
2        110       HIOS      21989AK0010001        AK         AK           6  
3        111       HIOS      21989AK0010001        AK         AK           6  
4        112       HIOS      21989AK0010001        AK         AK           6

In [47]:
# inspect the dataset
df_bcs2014.shape

(1164869, 32)

### 3. Select only the columns that you want to analyze
1. Select 'StandardComponentId','BenefitName','CopayInnTier1' columns.
2. Check 'CopayInnTier1' Column for to calculate copay value.

In [48]:
#Select only the columns you want to variable df_c1
df_c1 = df_bcs2014[['StandardComponentId','BenefitName','CopayInnTier1']]

In [49]:
## display all columns
pd.set_option('display.max_columns',None)
df_c1.head()

StandardComponentId                      BenefitName CopayInnTier1
0      21989AK0010001  Routine Dental Services (Adult)     No Charge
1      21989AK0010001     Dental Check-Up for Children     No Charge
2      21989AK0010001        Basic Dental Care - Child     No Charge
3      21989AK0010001              Orthodontia - Child     No Charge
4      21989AK0010001        Major Dental Care - Child     No Charge

In [50]:
# display all columns, Non-Null Count and Dtype
df_c1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1164869 entries, 0 to 1164868
Data columns (total 3 columns):
 #   Column               Non-Null Count    Dtype 
---  ------               --------------    ----- 
 0   StandardComponentId  1164869 non-null  object
 1   BenefitName          1164869 non-null  object
 2   CopayInnTier1        898756 non-null   object
dtypes: object(3)
memory usage: 35.5+ MB


In [51]:
# display first 5 rows
df_c1.head()

## display all rows
pd.set_option('display.max_rows',None)
df_c1['CopayInnTier1'].value_counts().reset_index()

CopayInnTier1   count
0                        No Charge  465861
1                               $0  201075
2       No Charge after deductible   54502
3                              $50   13936
4                 $0 Copay per Day   12833
5                              $20    9874
6                              $40    9359
7                              $30    9194
8                              $10    8919
9                              $25    8747
10               $0 Copay per Stay    7920
11                             $75    5755
12                             $60    4883
13                             $15    4824
14                             $35    4386
15       $0 Copay after deductible    3838
16                            $250    3180
17                            $100    3157
18      $40 Copay after deductible    2961
19      $50 Copay after deductible    2940
20                            $150    2519
21                            $500    2204
22                              $5    2198
23                             $45    1923
24      $30 Copay after deductible    1769
25                              0%    1742
26                            $200    1532
27      $75 Copay after deductible    1499
28     $100 Copay after deductible    1471
29    $250 Copay before deductible    1461
30      $10 Copay after deductible    1447
31                             $70    1384
32    $500 Copay before deductible    1307
33      $25 Copay after deductible    1261
34     $250 Copay after deductible    1257
35     $150 Copay after deductible    1174
36                             $55    1114
37     $20 Copay before deductible    1073
38      $20 Copay after deductible    1047
39    $100 Copay before deductible    1025
40     $75 Copay before deductible     984
41     $50 Copay before deductible     976
42     $10 Copay before deductible     952
43                             $80     821
44      $60 Copay after deductible     800
45      $70 Copay after deductible     781
46                              $8     769
47                              $4     753
48                            $300     723
49     $500 Copay after deductible     701
50     $30 Copay before deductible     691
51             $500 Copay per Stay     674
52     $25 Copay before deductible     650
53      $35 Copay after deductible     645
54      $15 Copay after deductible     625
55                             $65     622
56    $200 Copay before deductible     582
57      $45 Copay after deductible     542
58    $1000 Copay after deductible     537
59                              $3     510
60      $55 Copay after deductible     509
61     $60 Copay before deductible     456
62     $40 Copay before deductible     449
63             $250 Copay per Stay     438
64     $35 Copay before deductible     367
65      $65 Copay after deductible     355
66     $85 Copay before deductible     348
67                            $750     347
68     $200 Copay after deductible     346
69                             $90     336
70                            $400     321
71              $500 Copay per Day     309
72      $80 Copay after deductible     305
73                            $350     303
74                          $1,000     301
75    $110 Copay before deductible     294
76              $250 Copay per Day     278
77     $15 Copay before deductible     265
78       $3 Copay after deductible     247
79    $300 Copay before deductible     240
80     $300 Copay after deductible     232
81                              $2     221
82             $200 Copay per Stay     216
83                            $700     210
84                             $17     207
85     $45 Copay before deductible     204
86    $375 Copay before deductible     200
87    $150 Copay before deductible     188
88             $100 Copay per Stay     171
89     $125 Copay after deductible     155
90              $100 Copay per Day     155
91              $200 Copay per Day     151
92                    

### 4. Separate the string to find groups of string
1. Separate numbers from CopayInnTier1 and store it in Copayment
2. Remove the number part from CopayInnTier1 and keep only the description (Description)
3. Clean NaN or empty values.
4. Convert amount_in_usd provided as a number (float) for further analysis.

In [52]:
# Separate into 'Amount' and 'Description'
#df_c1['Copay'] = df_c1['CopayInnTier1'].str.extract(r'\$([\d,]+)')
#df_c1['Description'] = df_c1['CopayInnTier1'].str.replace(r'\$[\d,]+\s?', '', regex=True)

# Replace 'nan' or empty strings with NaN for clarity
#df_c1['Copay'] = df_c1['Copay'].str.replace(',', '').fillna('NaN')  # Remove commas for numerical conversion
#df_c1['Description'] = df_c1['Description'].replace({'': 'NaN', 'nan': 'NaN'})

In [53]:
# ดึงเฉพาะตัวเลข
df_c1['Copay'] = df_c1['CopayInnTier1'].str.extract(r'([\d,]+)').replace({',': ''}, regex=True)
# ดึงเฉพาะข้อความ (ตัดตัวเลขออก)
df_c1['Description'] = df_c1['CopayInnTier1'].str.replace(r'[\d,]+', '', regex=True).str.strip()

/var/folders/kc/wzqlm10x69xgkrpv4q3bp89w0000gn/T/ipykernel_8343/4054445701.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_c1['Copay'] = df_c1['CopayInnTier1'].str.extract(r'([\d,]+)').replace({',': ''}, regex=True)
/var/folders/kc/wzqlm10x69xgkrpv4q3bp89w0000gn/T/ipykernel_8343/4054445701.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_c1['Description'] = df_c1['CopayInnTier1'].str.replace(r'[\d,]+', '', regex=True).str.strip()


In [54]:
# display first 5 rows
df_c1.head()

StandardComponentId                      BenefitName CopayInnTier1 Copay  \
0      21989AK0010001  Routine Dental Services (Adult)     No Charge   NaN   
1      21989AK0010001     Dental Check-Up for Children     No Charge   NaN   
2      21989AK0010001        Basic Dental Care - Child     No Charge   NaN   
3      21989AK0010001              Orthodontia - Child     No Charge   NaN   
4      21989AK0010001        Major Dental Care - Child     No Charge   NaN   

  Description  
0   No Charge  
1   No Charge  
2   No Charge  
3   No Charge  
4   No Charge

### 5. Information used in Two-way ANOVA
The dependent variable must be quantitative in order to be able to analyze differences in means between groups. Therefore, the data in the example you provide must be adjusted or manipulated to provide a suitable numerical value for analysis.

#### 5.1 Understanding Copay Information
From the information you have:
1. No Charge and No Charge after deductible: indicates that there is no charge. or not waste money
2. NaN: May mean missing or unknown data.
3. `$10` Copay after deductible, `$10` Copay before deductible, `$10` Copay per Day, `$10` Copay per Stay: should be converted to numbers that can be analyzed statistically.

#### 5.2 Handle non-numeric values
1. `$10, $500, $500`: Numbers are values that should be converted to numbers (0.10, 250, or 500, etc.).
2. No Charge / No Charge after deductible: should be converted to 0.
3. NaN: Show missing information. It can be drop because we don't domain expert.
4. 0%: Should be converted to 0.
5. Text formats such as after deductible or per Day should be separated or omitted if they are not relevant to the numbers.

#### Copay after deductible, Copay per Day, Copay before deductible, Copay per Stay. It has no effect because in the end you will have to pay the displayed amount anyway. It can be drop.

In [55]:
# Check Desciption values
value_counts = df_c1['Description'].value_counts().reset_index(name='Count')
value_counts

Description   Count
0                   No Charge  465861
1                           $  309579
2  No Charge after deductible   54502
3    $ Copay after deductible   28789
4             $ Copay per Day   14754
5   $ Copay before deductible   13662
6            $ Copay per Stay    9867
7                           %    1742

In [56]:
df_c1['Description'].unique()

array(['No Charge', nan, '$ Copay per Day', '$',
       '$ Copay before deductible', '$ Copay after deductible',
       '$ Copay per Stay', 'No Charge after deductible', '%'],
      dtype=object)

In [57]:
df_c1['Copay'].unique()

array([nan, '0', '45', '15', '10', '25', '30', '35', '50', '60', '17',
       '8', '20', '150', '100', '200', '40', '300', '55', '395', '75',
       '5', '70', '80', '65', '500', '350', '1000', '750', '125', '85',
       '250', '3', '90', '160', '120', '95', '110', '180', '1500', '1250',
       '400', '1400', '72', '800', '600', '9', '6', '4000', '4', '2', '1',
       '2500', '475', '54', '44', '430', '2800', '375', '16', '875',
       '275', '2200', '12', '46', '1695', '185', '2410', '335', '28',
       '2000', '320', '700', '310', '2600', '2700', '49', '7', '21', '33',
       '2250', '3000', '4500', '2100', '31', '105', '190', '130', '950',
       '162', '143', '22', '272', '226', '131', '418', '119', '109',
       '270', '499', '89', '74', '154', '87', '159', '38', '76', '3450',
       '175', '450', '63', '3540', '57', '18', '280', '48', '26', '145',
       '66', '225', '140', '1200', '205', '144', '449', '589', '279',
       '427', '155', '550', '650', '825', '88', '210', '900', '4

### 6. Handle this data
1. Replace No Charge, No Charge after deductible, 0% in Description columns = 0 in Copayment columns
2. Drop NaN in Decriptions columns

In [58]:
# Remove NaN 309579 records from 898756 = 589177
# Step1: Replace 'No Charge', 'No Charge after deductible', '0%' in Description to Copayment is 0
df_c1.loc[df_c1['Description'].isin(['No Charge', 'No Charge after deductible', '%']), 'Copay'] = '0'

# Step2: Drop 'NaN' in Copayment columns
# df_c1 = df_c1[df_c1['Description'] != 'NaN'].reset_index(drop=True)
df_cleaned = df_c1.dropna()

In [59]:
df_cleaned['Copay'] = pd.to_numeric(df_cleaned['Copay'], errors='coerce')

/var/folders/kc/wzqlm10x69xgkrpv4q3bp89w0000gn/T/ipykernel_8343/4145348928.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['Copay'] = pd.to_numeric(df_cleaned['Copay'], errors='coerce')


In [60]:
df_cleaned['Copay'].unique()

array([   0,   45,   15,   10,   25,   30,   35,   50,   60,   17,    8,
         20,  150,  100,  200,   40,  300,   55,  395,   75,    5,   70,
         80,   65,  500,  350, 1000,  750,  125,   85,  250,    3,   90,
        160,  120,   95,  110,  180, 1500, 1250,  400, 1400,   72,  800,
        600,    9,    6, 4000,    4,    2,    1, 2500,  475,   54,   44,
        430, 2800,  375,   16,  875,  275, 2200,   12,   46, 1695,  185,
       2410,  335,   28, 2000,  320,  700,  310, 2600, 2700,   49,    7,
         21,   33, 2250, 3000, 4500, 2100,   31,  105,  190,  130,  950,
        162,  143,   22,  272,  226,  131,  418,  119,  109,  270,  499,
         89,   74,  154,   87,  159,   38,   76, 3450,  175,  450,   63,
       3540,   57,   18,  280,   48,   26,  145,   66,  225,  140, 1200,
        205,  144,  449,  589,  279,  427,  155,  550,  650,  825,   88,
        210,  900,   43,  196,   32,   96,  480,  578,  465,   81,   79,
         69])

In [61]:
df_cleaned['Description'].unique()

array(['No Charge', '$ Copay per Day', '$', '$ Copay before deductible',
       '$ Copay after deductible', '$ Copay per Stay',
       'No Charge after deductible', '%'], dtype=object)

In [62]:
df_cleaned.head()

StandardComponentId                      BenefitName CopayInnTier1  Copay  \
0      21989AK0010001  Routine Dental Services (Adult)     No Charge      0   
1      21989AK0010001     Dental Check-Up for Children     No Charge      0   
2      21989AK0010001        Basic Dental Care - Child     No Charge      0   
3      21989AK0010001              Orthodontia - Child     No Charge      0   
4      21989AK0010001        Major Dental Care - Child     No Charge      0   

  Description  
0   No Charge  
1   No Charge  
2   No Charge  
3   No Charge  
4   No Charge

In [63]:
# display all columns, Non-Null Count and Dtype
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 898756 entries, 0 to 1164864
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   StandardComponentId  898756 non-null  object
 1   BenefitName          898756 non-null  object
 2   CopayInnTier1        898756 non-null  object
 3   Copay                898756 non-null  int64 
 4   Description          898756 non-null  object
dtypes: int64(1), object(4)
memory usage: 41.1+ MB


### 7. Find out Mean of Copay
1. Find out Mean of Copayment each PlanId and Add new columns Mean Copayment for check each PlanId have how mean of Copayment
2. Check manual
3. Calculate the Mean of each PlanId and take the resulting Mean file and add it as a new Column after the Description column by inserting the reference value from the original PlanId
4. Check Mean_Copayment of StandardComponentId that you want

In [64]:
# Step1: Find out Copayment > 0 for test find out mean value of PlanId
plans_with_copayment = df_cleaned[df_cleaned['Copay'] > 0]['StandardComponentId'].unique()
plans_with_copayment

array(['73836AK0620001', '73836AK0620002', '73836AK0630001', ...,
       '31274WV0350002', '31274WV0420021', '31274WV0420031'], dtype=object)

In [65]:
# Step2: Check manual
# Filtered PlanId is 31274WV0420031 (Example)
filtered_df = df_cleaned[df_cleaned['StandardComponentId'] == '31274WV0420031']

# Count all value of Copayment
copayment_count = filtered_df['Copay'].notna().sum()

# Find valure that not duplicate value in Copayment and number of value
copayment_value_counts = filtered_df['Copay'].value_counts().reset_index()
copayment_value_counts.columns = ['Copay', 'Count']

# Calculate Mean of Copayment
copayment_mean = filtered_df['Copay'].mean()

# Display recheck mean value of 31274WV0420031
print(f"all Copayment values: {copayment_count}")
print("Values in Copay:")
print(copayment_value_counts)
print(f"Mean of Copayment: {copayment_mean:.2f}")

#Check manual 31274WV0420031 = (35+35+35+35+25+25+40+40+4+70)/62 = 5.55 Mean

all Copayment values: 62
Values in Copay:
   Copay  Count
0      0     52
1     35      4
2     25      2
3     40      2
4      4      1
5     70      1
Mean of Copayment: 5.55


In [66]:
# Step3: Calculate Mean of Copayment for each PlanId
mean_copayment_per_plan = df_cleaned.groupby('StandardComponentId')['Copay'].mean().reset_index(name='Copayment')

# Merge the Mean value back into the original DataFrame
df_with_mean = pd.merge(df_cleaned, mean_copayment_per_plan, on='StandardComponentId', how='left')

# Reorder columns to be after Description
columns_order = df_cleaned.columns.tolist() + ['Copayment']
df_with_mean = df_with_mean[columns_order]

# Display first 5 rows
df_with_mean.head()

StandardComponentId                      BenefitName CopayInnTier1  Copay  \
0      21989AK0010001  Routine Dental Services (Adult)     No Charge      0   
1      21989AK0010001     Dental Check-Up for Children     No Charge      0   
2      21989AK0010001        Basic Dental Care - Child     No Charge      0   
3      21989AK0010001              Orthodontia - Child     No Charge      0   
4      21989AK0010001        Major Dental Care - Child     No Charge      0   

  Description  Copayment  
0   No Charge        0.0  
1   No Charge        0.0  
2   No Charge        0.0  
3   No Charge        0.0  
4   No Charge        0.0

In [67]:
# display all columns, Non-Null Count and Dtype
df_with_mean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 898756 entries, 0 to 898755
Data columns (total 6 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   StandardComponentId  898756 non-null  object 
 1   BenefitName          898756 non-null  object 
 2   CopayInnTier1        898756 non-null  object 
 3   Copay                898756 non-null  int64  
 4   Description          898756 non-null  object 
 5   Copayment            898756 non-null  float64
dtypes: float64(1), int64(1), object(4)
memory usage: 41.1+ MB


In [68]:
# Step4: Check Copayment of PlanId that you want
mean_value = df_with_mean.loc[df_with_mean['StandardComponentId'] == '31274WV0420031', 'Copayment'].iloc[0]

# Display
print(f"Copayment value of PlanId 31274WV0420031 is: {mean_value:.2f}")

Copayment value of PlanId 31274WV0420031 is: 5.55


In [69]:
# Filter Particular PlanId that you want "31274WV0420031"
filtered_df = df_with_mean.loc[df_with_mean['StandardComponentId'] == '31274WV0420031']

# Display all rows
filtered_df

StandardComponentId                                        BenefitName  \
897932      31274WV0420031                                          Radiation   
897933      31274WV0420031                                 Diabetes Education   
897934      31274WV0420031                                 Prosthetic Devices   
897935      31274WV0420031                                   Infusion Therapy   
897936      31274WV0420031    Treatment for Temporomandibular Joint Disorders   
897937      31274WV0420031                             Reconstructive Surgery   
897938      31274WV0420031                                    Clinical Trials   
897939      31274WV0420031                           Diabetes Care Management   
897940      31274WV0420031                                  Dental Anesthesia   
897941      31274WV0420031                                Mental Health Other   
897942      31274WV0420031                           Prescription Drugs Other   
898259      31274WV0420031   Primary Care Visit to Treat an Injury or Illness   
898260      31274WV0420031                                   Specialist Visit   
898261      31274WV0420031  Other Practitioner Office Visit (Nurse, Physic...   
898262      31274WV0420031  Outpatient Facility Fee (e.g.,  Ambulatory Sur...   
898263      31274WV0420031     Outpatient Surgery Physician/Surgical Services   
898264      31274WV0420031                                   Hospice Services   
898265      31274WV0420031  Non-Emergency Care When Traveling Outside the ...   
898266      31274WV0420031                               Private-Duty Nursing   
898267      31274WV0420031                           Routine Eye Exam (Adult)   
898268      31274WV0420031                  Urgent Care Centers or Facilities   
898269      31274WV0420031                          Home Health Care Services   
898270      31274WV0420031                            Emergency Room Services   
898271      31274WV0420031                 Emergency Transportation/Ambulance   
898272      31274WV0420031  Inpatient Hospital Services (e.g., Hospital Stay)   
898273      31274WV0420031          Inpatient Physician and Surgical Services   
898274      31274WV0420031                                  Bariatric Surgery   
898275      31274WV0420031                                   Cosmetic Surgery   
898276      31274WV0420031                           Skilled Nursing Facility   
898277      31274WV0420031                        Prenatal and Postnatal Care   
898278      31274WV0420031  Delivery and All Inpatient Services for Matern...   
898279      31274WV0420031       Mental/Behavioral Health Outpatient Services   
898280      31274WV0420031        Mental/Behavioral Health Inpatient Services   
898281      31274WV0420031       Substance Abuse Disorder Outpatient Services   
898282      31274WV0420031        Substance Abuse Disorder Inpatient Services   
898283      31274WV0420031                                      Generic Drugs   
898284      31274WV0420031                              Preferred Brand Drugs   
898285      31274WV0420031                          Non-Preferred Brand Drugs   
898286      31274WV0420031                                    Specialty Drugs   
898287      31274WV0420031                 Outpatient Rehabilitation Services   
898288      31274WV0420031                              Habilitation Services   
898289      31274WV0420031                                  Chiropractic Care   
898290      31274WV0420031                          Durable Medical Equipment   
898291      31274WV0420031                       Imaging (CT/PET Scans, MRIs)   
898292      31274WV0420031             Preventive Care/Screening/Immunization   
898293      31274WV0420031                      Routine Eye Exam for Children   
898294      31274WV0420031                           Eye Glasses for Children   
898295      31274WV0420031                       Dental Check-Up for Children   
898296      31274WV0420031            

In [72]:
df_with_mean['Copayment'] = np.ceil(df_with_mean['Copayment']).astype(int)

In [73]:
# Filter Particular PlanId that you want "31274WV0420031"
filtered_df = df_with_mean.loc[df_with_mean['StandardComponentId'] == '31274WV0420031']

# Display all rows
filtered_df

StandardComponentId                                        BenefitName  \
897932      31274WV0420031                                          Radiation   
897933      31274WV0420031                                 Diabetes Education   
897934      31274WV0420031                                 Prosthetic Devices   
897935      31274WV0420031                                   Infusion Therapy   
897936      31274WV0420031    Treatment for Temporomandibular Joint Disorders   
897937      31274WV0420031                             Reconstructive Surgery   
897938      31274WV0420031                                    Clinical Trials   
897939      31274WV0420031                           Diabetes Care Management   
897940      31274WV0420031                                  Dental Anesthesia   
897941      31274WV0420031                                Mental Health Other   
897942      31274WV0420031                           Prescription Drugs Other   
898259      31274WV0420031   Primary Care Visit to Treat an Injury or Illness   
898260      31274WV0420031                                   Specialist Visit   
898261      31274WV0420031  Other Practitioner Office Visit (Nurse, Physic...   
898262      31274WV0420031  Outpatient Facility Fee (e.g.,  Ambulatory Sur...   
898263      31274WV0420031     Outpatient Surgery Physician/Surgical Services   
898264      31274WV0420031                                   Hospice Services   
898265      31274WV0420031  Non-Emergency Care When Traveling Outside the ...   
898266      31274WV0420031                               Private-Duty Nursing   
898267      31274WV0420031                           Routine Eye Exam (Adult)   
898268      31274WV0420031                  Urgent Care Centers or Facilities   
898269      31274WV0420031                          Home Health Care Services   
898270      31274WV0420031                            Emergency Room Services   
898271      31274WV0420031                 Emergency Transportation/Ambulance   
898272      31274WV0420031  Inpatient Hospital Services (e.g., Hospital Stay)   
898273      31274WV0420031          Inpatient Physician and Surgical Services   
898274      31274WV0420031                                  Bariatric Surgery   
898275      31274WV0420031                                   Cosmetic Surgery   
898276      31274WV0420031                           Skilled Nursing Facility   
898277      31274WV0420031                        Prenatal and Postnatal Care   
898278      31274WV0420031  Delivery and All Inpatient Services for Matern...   
898279      31274WV0420031       Mental/Behavioral Health Outpatient Services   
898280      31274WV0420031        Mental/Behavioral Health Inpatient Services   
898281      31274WV0420031       Substance Abuse Disorder Outpatient Services   
898282      31274WV0420031        Substance Abuse Disorder Inpatient Services   
898283      31274WV0420031                                      Generic Drugs   
898284      31274WV0420031                              Preferred Brand Drugs   
898285      31274WV0420031                          Non-Preferred Brand Drugs   
898286      31274WV0420031                                    Specialty Drugs   
898287      31274WV0420031                 Outpatient Rehabilitation Services   
898288      31274WV0420031                              Habilitation Services   
898289      31274WV0420031                                  Chiropractic Care   
898290      31274WV0420031                          Durable Medical Equipment   
898291      31274WV0420031                       Imaging (CT/PET Scans, MRIs)   
898292      31274WV0420031             Preventive Care/Screening/Immunization   
898293      31274WV0420031                      Routine Eye Exam for Children   
898294      31274WV0420031                           Eye Glasses for Children   
898295      31274WV0420031                       Dental Check-Up for Children   
898296      31274WV0420031            

### 8. Drop Duplicates values
1. Drop 'StandardComponentId' column with duplicate values
2. Create CSV file

In [74]:
#drop duplicates values keep in df_no_duplicates
df_no_duplicates = df_with_mean.drop_duplicates(subset=["StandardComponentId"])
df_no_duplicates

StandardComponentId                                        BenefitName  \
0           21989AK0010001                    Routine Dental Services (Adult)   
14          21989AK0020001                    Routine Dental Services (Adult)   
28          21989AK0020002                    Routine Dental Services (Adult)   
42          38536AK0010002                          Major Dental Care - Adult   
45          38536AK0010003                    Routine Dental Services (Adult)   
50          73836AK0620002  Inpatient Hospital Services (e.g., Hospital Stay)   
58          42507AK0030001        Non-Medically Necessary Orthodontia - Child   
59          42507AK0040001                    Routine Dental Services (Adult)   
73          42507AK0050001                    Routine Dental Services (Adult)   
89          42507AK0060001                    Routine Dental Services (Adult)   
103         45858AK0030001                       Dental Check-Up for Children   
106         38536AK0010001                    Routine Dental Services (Adult)   
128         42507AK0010001                       Dental Check-Up for Children   
136         42507AK0020001                       Dental Check-Up for Children   
159         45858AK0030002                       Dental Check-Up for Children   
162         45858AK0040001                       Dental Check-Up for Children   
165         45858AK0040002                       Dental Check-Up for Children   
168         73836AK0620001   Primary Care Visit to Treat an Injury or Illness   
260         73836AK0630001   Primary Care Visit to Treat an Injury or Illness   
488         73836AK0630002   Primary Care Visit to Treat an Injury or Illness   
628         73836AK0630003   Primary Care Visit to Treat an Injury or Illness   
1115        73836AK0630004   Primary Care Visit to Treat an Injury or Illness   
1514        73836AK0640001   Primary Care Visit to Treat an Injury or Illness   
1714        73836AK0640002   Primary Care Visit to Treat an Injury or Illness   
1914        73836AK0650001   Primary Care Visit to Treat an Injury or Illness   
2174        73836AK0650002                      Rehabilitative Speech Therapy   
2464        73836AK0650003   Primary Care Visit to Treat an Injury or Illness   
2640        73836AK0650004   Primary Care Visit to Treat an Injury or Illness   
3014        73836AK0680004   Primary Care Visit to Treat an Injury or Illness   
3116        73836AK0680007   Primary Care Visit to Treat an Injury or Illness   
3218        73836AK0680011   Primary Care Visit to Treat an Injury or Illness   
3320        73836AK0680015   Primary Care Visit to Treat an Injury or Illness   
3422        73836AK0680017   Primary Care Visit to Treat an Injury or Illness   
3524        73836AK0680019   Primary Care Visit to Treat an Injury or Illness   
3626        73836AK0680027   Primary Care Visit to Treat an Injury or Illness   
3728        73836AK0680030   Primary Care Visit to Treat an Injury or Illness   
3830        73836AK0680034   Primary Care Visit to Treat an Injury or Illness   
3887        73836AK0680040                          Durable Medical Equipment   
3918        73836AK0680038                  Urgent Care Centers or Facilities   
4092        74819AK0010004                          Basic Dental Care - Adult   
4101        74819AK0010005                    Routine Dental Services (Adult)   
4115        74819AK0010006                    Routine Dental Services (Adult)   
4129        74819AK0020001                       Dental Check-Up for Children   
4133        74819AK0020004                          Basic Dental Care - Child   
4138        74819AK0020005                    Routine Dental Services (Adult)   
4144        74819AK0020002                       Dental Check-Up for Children   
4152        74819AK0020003                    Routine Dental Services (Adult)   
4155        73836AK0680042                 Inherited Metabolic Disorder - PKU   
4321        74819AK0010001            

In [75]:
# display all columns, Non-Null Count and Dtype
df_no_duplicates.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6632 entries, 0 to 898748
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   StandardComponentId  6632 non-null   object
 1   BenefitName          6632 non-null   object
 2   CopayInnTier1        6632 non-null   object
 3   Copay                6632 non-null   int64 
 4   Description          6632 non-null   object
 5   Copayment            6632 non-null   int64 
dtypes: int64(2), object(4)
memory usage: 362.7+ KB


In [76]:
# display first 5 rows
df_no_duplicates.head()

StandardComponentId                      BenefitName CopayInnTier1  Copay  \
0       21989AK0010001  Routine Dental Services (Adult)     No Charge      0   
14      21989AK0020001  Routine Dental Services (Adult)     No Charge      0   
28      21989AK0020002  Routine Dental Services (Adult)     No Charge      0   
42      38536AK0010002        Major Dental Care - Adult     No Charge      0   
45      38536AK0010003  Routine Dental Services (Adult)     No Charge      0   

   Description  Copayment  
0    No Charge          0  
14   No Charge          0  
28   No Charge          0  
42   No Charge          0  
45   No Charge          0

In [77]:
df_no_duplicates['Copayment'].unique()

array([  0,   1,   2,   4,   8,  12,   5,   3,  11,  14,  29,  25,  22,
        33,  46,   6,  31,  10,   9,  69,  26, 141, 140,  13,  18,  19,
        17,  23,  24,  32,  16,  15,  20, 127, 162,  30,  37,  42,  70,
        74, 116, 128, 205,  99,  94, 199, 175, 278,  72,  45,  85,  98,
        92,  68,  40,  73,  81,  97,  71,  76, 104, 123,  61, 100, 131,
       110,  27, 129,  55, 113, 138, 650, 656, 781, 778, 757,  39,  21,
         7,  35, 117, 135, 172, 558, 180, 327, 447,  75, 163,  83,  62,
        57,  91,  84, 126,  60, 157, 159,  87,  82, 275, 263, 503, 523,
       524,  95,  89, 152,  86,  58, 213,  88, 722, 719, 700, 629, 668,
        28,  48,  49,  41,  52, 155, 211, 271, 280, 207, 416,  43,  44,
       647, 658,  59,  47,  51,  54,  63, 763, 750, 181, 436, 323,  77,
        66,  90, 101, 106, 165, 114, 124, 168, 246, 232, 137, 149, 191,
       221, 313, 315,  38, 256,  65,  50,  80, 108, 105,  34,  64,  36,
       148, 198, 526, 502, 164, 465, 136, 648, 654, 678, 146,  6

In [78]:
#create csv after clean BenefitCostSharing no duplicates
df_no_duplicates.to_csv('clean_benefitcostsharing_mean_no_duplicates.csv')